File: 6-2.rb
#!/usr/bin/env ruby

require 'set'

buffers = File.read('6.input').lines.map(&:strip)
buffers.each do |buf|
  buf.chars.each_index do |pos|
    next if pos < 13

    unique = buf[pos - 13..pos].chars.map { |s| Set.new(s.chars) }.reduce(&:|)
    if unique.size == 14
      print pos + 1, "\n"
      break
    end
  end
end

File: 6-2.py
#!/usr/bin/env python

from functools import reduce

buffers = [row.strip() for row in open('6.input').readlines()]

for buf in buffers:
    for pos in range(len(buf)):
        if pos < 13:
            continue

        test = buf[pos - 13:pos + 1]
        sets = [set(s) for s in test]
        unique = reduce(lambda a, b: a | b, sets)
        if len(unique) == 14:
            print(pos + 1)
            break
